server/pbenchinacan/load_keycloak.sh

```diff
@@ -13,12 +13,14 @@
 
 # The script defaults to using master realm username/password as admin/admin
 # unless specified otherwise by 'ADMIN_USERNAME' and 'ADMIN_PASSWORD' env
-# variables. The script also defaults the keycloak redirect URI as
-# "https://localhost:8443/*" unless specified otherwise by 'KEYCLOAK_REDIRECT_URI'
-# env variable.
+# variables. The script constructs the Keycloak redirect URI list as
+# "https://<node>:8443/*" from the space-separated list of nodes in
+# the 'KEYCLOAK_REDIRECT_HOSTS' env variable, defaulting to 'localhost' plus
+# whatever we can glean from the 'hostname' command.  It adds to that list the
+# value of 'KEYCLOAK_DEV_REDIRECT' which defaults to "http://localhost:3000/*".
 
 KEYCLOAK_HOST_PORT=${KEYCLOAK_HOST_PORT:-"https://localhost:8090"}
-KEYCLOAK_REDIRECT_URI=${KEYCLOAK_REDIRECT_URI:-"https://localhost:8443/*"}
+KEYCLOAK_REDIRECT_HOSTS=${KEYCLOAK_REDIRECT_HOSTS:-"localhost $(hostname -I) $(hostname -A) $(hostname -f)"}
```

Comment:
Dang: so `"localhost 10.0.0.104 <IPv6> dbutenho-thinkpadp1gen4i.bos.csb dbutenho-thinkpadp1gen4i.bos.csb dbutenho-thinkpadp1gen4i.bos.csb"` (that's my local router IPv4 since I'm not on VPN, and the csb FQDNs aren't routable, so the IPv6 is the only one I actually ought to avoid revealing on GitHub 😆 )

But, seriously, using `hostname -f` when you're using `hostname -A` seems pointless as the former is "the" FQDN while the latter is *all* FQDNs and therefore will always be a superset. (And, since I don't think any of our systems have separate FQDNs across NICs, probably *still* always a duplicate.)

Also, I'm not sure the redirect will take an undelimited IPv6 address, so your `https://<IPv6>:8443` probably won't work and *might* generate an error in Keycloak. I suspect it'd need to be `https://[<IPv6>]:8443` ... all of which makes this IP stuff a lot more complicated than it might seem at first.

I just think you're potentially asking for trouble here, and I'm not sure your result is really going to help anyone anyway since we're not using this in a real deployment anymore.

Reply:
OK, they are now wrapped.  Assuming the tests pass, I don't suppose you'd be willing to try the script on your machine?  (Mine machines don't speak IPv6 outside of containers, and my attempt to run `jenkins/runlocal` using `jenkins/run` built the RPM but then couldn't find it to build the container.... 🍤)

Reply:
> However, if your goal is to avoid duplication, I think you're out of luck -- I just got this:
> 
> ```
> # hostname -A
> 8e10ccf8e78e 8e10ccf8e78e 
> ```

Wait; so where did it get a "hostname" like that? Even inside a container, I'm seeing `host.containers.internal` -- which is stupid, but arguably less stupid than yours. 🤔 

> rofl
> 
> > since I don't think any of our systems have separate FQDNs across NICs, probably still always a duplicate.
> 
> `intlab-002` has two NICs, but the second one doesn't seem to have a FQDN (i.e., the reverse-lookup on its IPv4 address fails), but `hostname -A` there reports a duplication of the `n002` name (in addition to the `intlab-002` name), too.

Interesting. I know the BOS lab used to have a high-speed internal subnet; and presumably the "second" proxy NIC would be on the internal net, which I don't think had DNS. RDU2 may well do something similar.

> > I just think you're potentially asking for trouble here, and I'm not sure your result is really going to help anyone anyway since we're not using this in a real deployment anymore.
> 
> I may indeed be asking for trouble, since I don't seem to have a system with IPv6 configured that I can test on. However, the result helps _me_ -- running with the browser on my laptop and the canned server on my VM in the IntLab, the Keycloak redirect has to be something other than `localhost`. So, the question is, how to get that "something", and I was hoping that `hostname` could provide the answer...but there is no good way to choose among (or even _recognize_) what it offers. slightly_frowning_face

I really don't understand the logic of the CSB configuration. They *have* DHCP addresses, but apparently no DDNS names, and `hostname` generates only non-routable blather. Then again, VMs and containers generally aren't any better.

(Then again, on the staging VM, `hostname -f` gives "rhel9-pbenchstaging" while `hostname -A` gives the actual routable DDNS DHCP FQDN, so that's something.)

> > I'm not sure the redirect will take an undelimited IPv6 address
> 
> I just tried a manual test...and it did! woozy_face
> 
> ```
> "redirectUris": ["http://localhost:3000/*", "https://fd00::948d:1dff:fe6c:fefe:8443/*"]
> ```
> 
> So, I guess it just does a dumb-ish pattern match on the backside.

Leaving the obvious question of which form, if either, would *match* on a redirect.

> But, you're absolutely right, the IPv6 addresses should be wrapped in square-brackets.

I pulled your branch and did a `runlocal`. I verified that the Keycloak has the (delimited) IPv6 address. (Actually, it's got both the link-local address and the routable address ... and while my first thought was that the link-local was silly as it's *not* routable off the local subnet, well, my laptop curiously enough *is* on its own local subnet...) Anyway, I can open a web page on either delimited IPv6 address copied from the redirects list, and the page loads and I can log in!

And, interestingly, while the CSB `hostname` isn't actually routable, I'd never had occasion to notice before that it actually works as a "localhost" alias on my laptop, so `https://dbutenho-thinkpadp1gen4i.bos.csb:8443/dashboard/` loads (and logs in) just fine!

The upshot here is that every one of the redirect URLs you're programming on my laptop actually work to launch the dashboard and to log in: even the ridiculously silly ones. 😆

Reply:
Addendum: on a whim, I also tried the local dashboard dev-mode to exercise the `http://localhost:3000` redirect, although there was no reason to think it wouldn't still work. It does... but, feeling in an IPv6-ish mood, I tried connecting to IPv6 localhost `[::]`, and that's not mapped in your cert (or probably in the allowed redirects), so it didn't work. Hmm. But the link-local IPv6 (`PBENCH_SERVER=https://\[fd10:22:16:1::10b9\]:8443 npm run dev`) worked just fine.

Reply:
> > However, if your goal is to avoid duplication, I think you're out of luck -- I just got this:
> > ```
> > # hostname -A
> > 8e10ccf8e78e 8e10ccf8e78e 
> > ```
> 
> Wait; so where did it get a "hostname" like that?

No clue...I'm just reporting what I'm seeing.  🙃

> which is stupid, but arguably less stupid than yours. 🤔

😆

> presumably the "second" proxy NIC would be on the internal net, which I don't think had DNS. RDU2 may well do something similar.

I expect that you're correct.

> I really don't understand the logic of the CSB configuration. They _have_ DHCP addresses, but apparently no DDNS names, and `hostname` generates only non-routable blather. Then again, VMs and containers generally aren't any better.

Yeah, and it's not just CSB.  My home network has similarly awkward hostnames (which are not, generally, as far as I can tell, routable or even usable except in rare circumstances).

> (Then again, on the staging VM, `hostname -f` gives "rhel9-pbenchstaging" while `hostname -A` gives the actual routable DDNS DHCP FQDN, so that's something.)

Hah!  You found it!!  _That's_ a case of the return from `-f` not being in the list (of one...) returned by `-A`....

> > So, I guess it just does a dumb-ish pattern match on the backside.
> 
> Leaving the obvious question of which form, if either, would _match_ on a redirect.

Well, the redirect is provided by the client (to wit, the Dashboard) and checked by the OIDC server and _then_ invoked.  So, if the provided value doesn't match the valid-redirect list, it doesn't really matter whether it is a valid URL format.  🙃  And, if it does match the list, but isn't a valid URL, then that's a problem for the client...so, I'm not sure that Keycloak has any motivation for enforcing anything other than fairly simple pattern matching.

> I pulled your branch and did a `runlocal`. I verified that the Keycloak has the (delimited) IPv6 address. (Actually, it's got both the link-local address and the routable address ... and while my first thought was that the link-local was silly as it's _not_ routable off the local subnet, well, my laptop curiously enough _is_ on its own local subnet...) Anyway, I can open a web page on either delimited IPv6 address copied from the redirects list, and the page loads and I can log in!

Yay!  🎉   Thanks for checking that!!

> And, interestingly, while the CSB `hostname` isn't actually routable, I'd never had occasion to notice before that it actually works as a "localhost" alias on my laptop, so `https://dbutenho-thinkpadp1gen4i.bos.csb:8443/dashboard/` loads (and logs in) just fine!

I'm glad it's good for _something_....  😉 

> The upshot here is that every one of the redirect URLs you're programming on my laptop actually work to launch the dashboard and to log in: even the ridiculously silly ones. 😆

Hurrah!  (Thanks again for checking!!)

> I also tried the local dashboard dev-mode to exercise the `http://localhost:3000` redirect, although there was no reason to think it wouldn't still work. It does... but, feeling in an IPv6-ish mood, I tried connecting to IPv6 localhost `[::]`, and that's not mapped in your cert (or probably in the allowed redirects), so it didn't work. Hmm. But the link-local IPv6 (`PBENCH_SERVER=https://\[fd10:22:16:1::10b9\]:8443 npm run dev`) worked just fine.

OK...if you're willing to test it again, I'll see if I can close that gap....  🤔  But see [my comment below](https://github.com/distributed-system-analysis/pbench/pull/3503#discussion_r1282390177).